### PR TITLE
Add testing guide

### PR DESCRIPTION
**Pull request by akaszynski**
Initial attempt at adding testing section.

Will need to merge with new guidelines layout.

### CODE REVIEW

File: doc/source/documentation_style/deployment.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
+.. _api_documentation:
+
+
+Documentation Deployment
+------------------------
+Documentation for the PyAnsys project is hosted via `GitHub Pages <https://pages.github.com/>`_. Given the open-source nature of the project, documentation build by each PyAnsys library or project should be public and editable by users and the community at large.
```
Comment (PipKat) [inline doc/source/documentation_style/deployment.rst:6]:
```suggestion
Documentation for the PyAnsys project is hosted via `GitHub Pages <https://pages.github.com/>`_. 
Given the open-source nature of the project, documentation built by each PyAnsys library or
project should be public and editable by users and the community at large.
```

File: doc/source/documentation_style/index.rst [hunk 2]
```diff
@@ -29,10 +29,13 @@ The documentation for a PyAnsys library should contain:
 * User guide for the library.
 * Link to this developer's guide.
 
+Finally the documentation should be public and hosted via gh-pages, either as a branch named ``gh-pages`` within the library repository, or within a ``gh-pages`` branch within ``<library-repository>-docs``.  See :ref:`doc_deployment` for more details.
```
Comment (PipKat) [inline doc/source/documentation_style/index.rst:32]:
```suggestion
Finally the documentation should be public and hosted via gh-pages, either as a branch named ``gh-pages`` within the library repository or within a ``gh-pages`` branch within ``<library-repository>-docs``.  Fore more information,
see :ref:`doc_deployment`.
```

File: doc/source/overview/dev_practices.rst [hunk 3]
```diff
@@ -11,7 +11,7 @@ contributing directly to PyAnsys libraries.
 General Development Procedures
 ------------------------------
 
-To submit new code to a PyAnsys, first `fork <https://docs.github.com/en/get-started/quickstart/fork-a-repo>`_ 
+To submit new code to a PyAnsys, first `fork <https://docs.github.com/en/get-started/quickstart/fork-a-repo>`_
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:14]:
```suggestion
To submit new code to a PyAnsys library, first `fork <https://docs.github.com/en/get-started/quickstart/fork-a-repo>`_
```

File: doc/source/overview/dev_practices.rst [hunk 4]
```diff
@@ -11,7 +11,7 @@ contributing directly to PyAnsys libraries.
...
 the repository (for example `PyMAPDL <https://github.com/pyansys/pymapdl>`_)
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:15]:
```suggestion
the repository (for example, `PyMAPDL <https://github.com/pyansys/pymapdl>`_).
```

File: doc/source/overview/dev_practices.rst [hunk 5]
```diff
@@ -11,7 +11,7 @@ contributing directly to PyAnsys libraries.
...
 and then clone the forked repository to your local environment. Next, create a new branch based on the
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:16]:
```suggestion
Then, clone the forked repository to your local environment. Next, create a new branch based on the
```

File: doc/source/overview/dev_practices.rst (hunk 4 above)
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:15]:
```suggestion
the repository (for example, `PyMAPDL <https://github.com/pyansys/pymapdl>`_)
```

File: doc/source/overview/dev_practices.rst [hunk 6]
```diff
@@ -11,7 +11,7 @@ contributing directly to PyAnsys libraries.
...
 `Branch Naming Conventions Section <#branch-naming-conventions>`__ in
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:17]:
```suggestion
and then clone the forked repository to your local environment. Next, using our
`branch naming conventions <#branch-naming-conventions>`__, create a new branch in
```

File: doc/source/overview/dev_practices.rst [hunk 7]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
 
 Testing
 -------
-<Testing Section>
+Unit testing is critical for the sucessful continious integration and testing of
+any program or libraries belonging to the PyAnsys project are no exception.
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:198]:
```suggestion
any program or libraries belonging to the PyAnsys project.
```

File: doc/source/overview/dev_practices.rst [hunk 8]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+There are generally two types of libraries part of the PyAnsys project: those
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:200]:
```suggestion
The PyAnsys project generally has two types of libraries: those
```

Comment (PipKat) [reply doc/source/overview/dev_practices.rst:200]: These two might best be described in a bulleted list.

File: doc/source/overview/dev_practices.rst [hunk 9]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+that interface or wrap functionality of a different Ansys product, service, or
+application, or tools those that provide functionality. Both types of libraries
+should be tested, but the tests written will depend on purpose of the library.
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:203]:
```suggestion
should be tested, but the tests written will depend on the purpose of the library.
```

File: doc/source/overview/dev_practices.rst [hunk 10]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+For example, a library that is wrapping a gRPC interface would include tests of
+the gRPC methods exposed by the proto files and wrapped by the Python library, but would not be expected to test the functionality of the server.
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:205]:
```suggestion
the gRPC methods exposed by the proto files and wrapped by the Python library. They would not be expected to test the functionality of the server.
```

File: doc/source/overview/dev_practices.rst [hunk 11]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+     Node		node = 1;
+   }
+
+  service SomeService{
+
+     rpc GetNode(NodeRequest)				returns (NodeResponse);
+     // other methods
+   }
+
+Then youyour unit test would test the wrapped python function (for example,
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:233]:
```suggestion
Then your unit test would test the wrapped python function (for example,
```

File: doc/source/overview/dev_practices.rst [hunk 12]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+``get_node``).  For example the ``get_node`` method might be implemented as:
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:234]:
```suggestion
``get_node``).  You might implement the ``get_node`` method with:
```

File: doc/source/overview/dev_practices.rst [hunk 13]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+       >>> srv.create_node(1, 4.5, 9.0, 3.2)
+       >>> node = srv.get_node(1)
+       >>> node
+       (4.5, 9.0, 3.2)
+
+       """
+       resp = service_pb2.GetNode(index=index)
+       return resp.x, resp.y, resp.z
+
+Your test would be implemented (within ``tests/test_nodes.py``):
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:266]: It seems that we can remove the parentheses as this information seems essential.

File: doc/source/overview/dev_practices.rst [hunk 14]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+   def test_get_node(srv):
+       srv.clear()
+
+       node_index = 1
+       node_coord = 0, 10, 20
+       srv.create_node(node_index, node_coord*)
+       assert srv.get_node(node_index) == node_coord
+
+The goal of the unit test should be to test the wrapping of the interface rather
+than the product or service itself. In the case of the ``GetNode``, this method
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:279]:
```suggestion
than the product or service itself. In the case of ``GetNode``, this method
```

File: doc/source/overview/dev_practices.rst [hunk 15]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+should have already been tested when designing and developing the service.
+
+
+Testing Framework
+~~~~~~~~~~~~~~~~~
+For consistency, PyAnsys tools and libraries should use either the `unit test
+<some link>`_ or `pytest <some link>`_ frameworks for unit testing. As detailed
+in the :ref:`repo_dir_struct`, unit tests should be placed into the ``tests``
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:287]:
```suggestion
<some link>`_ or `pytest <some link>`_ frameworks for unit testing. As described
in :ref:`repo_dir_struct`, unit tests should be placed into the ``tests``
```

File: doc/source/overview/dev_practices.rst [hunk 16]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+directory in the of the root directory of the library::
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:288]:
```suggestion
directory in the root directory of the library::
```

File: doc/source/overview/dev_practices.rst [hunk 17]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+   tests/
+       test_basic.py
+       test_advanced.py
+
+Furthermore, any testing dependencies requirements should be included when using ``setup.py`` within a ``requirements_tests.txt`` file that is installed via::
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:294]:
```suggestion
Furthermore, any requirements for testing dependencies should be included when using ``setup.py`` within a ``requirements_tests.txt`` file that is installed via::
```

File: doc/source/overview/dev_practices.rst [hunk 18]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+.. code::
+
+   pip install -r requirements_tests.txt
+
+Or included in ``pyproject.toml``. For example, when using the `filt
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:300]:
```suggestion
An alternative is to include requirements for dependencie in ``pyproject.toml``. For example, when using the `filt
```

File: doc/source/overview/dev_practices.rst [hunk 19]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+<missing_link>`_ build system::
+
+   [project.optional-dependencies]
+   test = [
+       "pytest>=2.7.3",
+       "pytest-cov",
+   ]
+
+And then installed via::
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:309]: We should format this as a complete sentence. Are the dependencies installed via the code presented next?

File: doc/source/overview/dev_practices.rst [hunk 20]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+   pip install .[test]
+
+When using ``pytest``, test via::
+
+   pytest
+
+.. note::
+   Note that it is preferable to cd into the testing directory and run unit
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:318]:
```suggestion
   We recommend that you use ``cd`` to change to the ``testing`` directory and run unit
```

File: doc/source/overview/dev_practices.rst [hunk 21]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+   testing there as you will be testing the installed library (generally in
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:319]:
```suggestion
   testing there because you will be testing the installed library (generally in
```

File: doc/source/overview/dev_practices.rst [hunk 22]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+   development mode ``pip install -e .``) rather than the source within the
+   uninstalled "local" source. This catches files that might be missed by the
+   installer, including any C extensions or additional internal packages.
+
+
+Coverage
+~~~~~~~~
+As Python is an interpreted language, developers of any Python library should
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:327]:
```suggestion
Because Python is an interpreted language, developers of a Python library should
```

File: doc/source/overview/dev_practices.rst [hunk 23]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+aim to have high coverage for their library as only syntax errors can be caught
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:328]:
```suggestion
aim to have high coverage for their libraries as only syntax errors can be caught
```

File: doc/source/overview/dev_practices.rst [hunk 24]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+during the almost trivial compile time. Coverage is defined as parts of the
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:329]:
```suggestion
during the almost trivial compile times. Coverage is defined as parts of the
```

File: doc/source/overview/dev_practices.rst [hunk 25]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+executable and usable source that are tested by unit tests, and the coverage for
+your library can be viewed using the ``pytest-cov`` library. For example::
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:331]:
```suggestion
executable and usable source that are tested by unit tests. You can use the ``pytest-cov`` library
to view the coverage for your library. For example::
```

File: doc/source/overview/dev_practices.rst [hunk 26]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+   Name                               Stmts   Miss  Cover
+   ------------------------------------------------------
+   numpydoc_validation/__init__.py        2      0   100%
+   numpydoc_validation/_validate.py      69      0   100%
+   ------------------------------------------------------
+   TOTAL                                 71      0   100%
+
+While 100% coverage is ideal, the law of diminishing returns often applies to
+the coverage of a Python library, and it is often sufficient, to achieve 80-90%
+coverage. For parts of your library that are difficult or impossible to test,
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:352]:
```suggestion
the coverage of a Python library. Consequently, achieving 80-90% coverage is often sufficient.
For parts of your library that are difficult or impossible to test,
```

File: doc/source/overview/dev_practices.rst [hunk 27]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+consider using ``# pragma: no cover`` at the end of method definition, branch,
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:353]:
```suggestion
consider using ``# pragma: no cover`` at the end of the method definition, branch,
```

File: doc/source/overview/dev_practices.rst [hunk 28]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+example, if part of your module performs a simple ``import`` test of
+``matplotlib`` and raises an error when the library is not installed, it is not
+reasonable to attempt to test this and assume full coverage:
+
+.. code:: python
+
+   try:
+       import matplotlib
+   except ImportError:  # pragma: no cover
+       raise ImportError("Please install matplotlib to use this feature")
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:364]:
```suggestion
       raise ImportError("Install matplotlib to use this feature.")
```

File: doc/source/overview/dev_practices.rst [hunk 29]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+.. note::
+   You should only avoid coverage of parts of your library where you cannot
+   reasonably test without an extensive testing suite or setup.  Most methods
+   and classes, including edge cases, can be reasonable tested. Even parts of
+   your code that raise errors like ``TypeError`` or ``ValueError`` when users
+   input the wrong data type or value.
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:371]:
```suggestion
   input the wrong data type or value can be reasonably tested.
```

File: doc/source/overview/dev_practices.rst [hunk 30]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+
+Unit Testing within CI/CD
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:374]:
```suggestion
Unit Testing Within CI/CD
```

File: doc/source/overview/dev_practices.rst [hunk 31]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+~~~~~~~~~~~~~~~~~~~~~~~~~
+Effective CI/CD assumes that unit testing is developed during feature
+development or bug fixes. However, given the limited scope of the local
+development environment, it is often not possible to enforce testing on multiple
+platforms, or even, unit testing in general. However, with the right automated
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:379]:
```suggestion
platforms, or even unit testing in general. However, with the right automated
```

File: doc/source/overview/dev_practices.rst [hunk 32]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+CI/CD, such testing can still occur and be enforced automatically.
+
+`GitHub Actions <gh actions link>`_ is the preferred automated CI/CD platform
+for running Python library unit tests for PyAnsys, and can be employed
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:383]:
```suggestion
for running Python library unit tests for PyAnsys. You can employ it
```

File: doc/source/overview/dev_practices.rst [hunk 33]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+immediately by closing the project `template <link to
+github.com/pyansys/template>`_. For a basic overview of those unfamiliar of
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:385]:
```suggestion
github.com/pyansys/template>`_. If you are unfamiliar with
```

File: doc/source/overview/dev_practices.rst [hunk 34]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+GitHub Actions, read: `missing link <missing_link>`_.
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:386]:
```suggestion
github.com/pyansys/template>`_. If you are unfamiliar with GitHub Actions, see: `missing link <missing_link>`_
for an overview.
```

File: doc/source/overview/dev_practices.rst [hunk 35]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+**Sample Workflow**
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:388]: If we are referring to the following sections, this should be a heading with the following entries as subheadings.

File: doc/source/overview/dev_practices.rst [hunk 36]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+The following sections detail the usage of a simple GitHub workflow for a
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:390]:
```suggestion
The following sections describe the usage of a simple GitHub workflow for a
```

File: doc/source/overview/dev_practices.rst [hunk 37]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+PyAnsys library:
+
+**Setup**
+
+Include the name and when your job should be run at the top of the workflow ``.yml``::
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:395]:
```suggestion
Include the unit test name and when it should be run at the top of the workflow ``.yml``::
```

File: doc/source/overview/dev_practices.rst [hunk 38]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+   on:
+     pull_request:
+     workflow_dispatch:
+     push:
+       tags:
+         - "*"
+       branches:
+         - main
+
+Take note that here we run the this workflow on all pull requests and on demand
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:408]:
```suggestion
Take note that this workflow runs on all pull requests and on demand
```

File: doc/source/overview/dev_practices.rst [hunk 39]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+with ``workflow_dispatch``, but limit it being run on commits to only tags and
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:409]:
```suggestion
with ``workflow_dispatch``. On commits, this workflow runs only on tags and
```

File: doc/source/overview/dev_practices.rst [hunk 40]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+on the ``main`` branch.  This ensures that CI/CD is not run twice on every
+commit for each PR, which may saturate available build or testing machines.
+
+**Job Description**
+
+PyAnsys Libraries should run on the currently supported versions of Python on both Windows and Linux (and ideally Mac OS). Therefore, it is necessary to also test on both Linux and Windows for these versions of Python. Use the ``matrix`` run strategy for the job with both the latest images of Windows and Linux::
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:415]:
```suggestion
PyAnsys libraries should run on the currently supported versions of Python on both Windows and Linux (and ideally on Mac OS). Therefore, it is necessary to also test on both Linux and Windows for these versions of Python. Use the ``matrix`` run strategy for the job with both the latest images of Windows and Linux::
```

File: doc/source/overview/dev_practices.rst [hunk 41]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+       name: Unit testing
+       runs-on: ${{ matrix.os }}
+       strategy:
+         matrix:
+           os: [windows-latest, ubuntu-latest]
+           python-version: ['3.7', '3.8', '3.9', '3.10']
+
+**Running the Tests**
+
+Each virtual machine within GitHub actions starts in a fresh state with no
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:428]: Should it be GitHub Actions?

File: doc/source/overview/dev_practices.rst [hunk 42]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+software or source installed or downloaded. Therefore it is necessary to both
+clone the repository using the ``checkout`` action, as well as setup Python and
+install the necessary testing dependencies.
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:431]:
```suggestion
software or source installed or downloaded. Therefore, you must clone the repository using the ``checkout`` action, set up Python, and install the necessary testing dependencies.
```

File: doc/source/overview/dev_practices.rst [hunk 43]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v1
+        with:
+          python-version: ${{ matrix.python-version }}
+
+If you are using ``setup.py``, your installation step will be::
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:441]:
```suggestion
If you are using ``setup.py``, your installation step is::
```

File: doc/source/overview/dev_practices.rst [hunk 44]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+      - name: Install the library
+        run: |
+          pip install .
+          pip install -r requirements_test.txt
+
+If you are using ``pyproject.toml`` with the ``flit`` build system, your
+installation step will be::
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:449]:
```suggestion
installation step is::
```

File: doc/source/overview/dev_practices.rst [hunk 45]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+          flit install
+
+Run the unit tests via ``pytest`` with::
+
+      - name: Test and show coverage
+        working-directory: tests
+        run: pytest --cov ansys.product.library
+
+.. note::
+   Replace ``ansys.product.library`` with your library. This should match how it
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:463]:
```suggestion
   Replace ``ansys.product.library`` with your library name. This should match how it
```

File: doc/source/overview/dev_practices.rst [hunk 46]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+           if_no_uploads: error
+       patch:
+         default:
+           # basic
+           target: 90%
+           if_not_found: success
+           if_ci_failed: error
+           if_no_uploads: error
+
+This requires that each PR has a patch coverage of 90%, meaning that 90% any source added to the repository (unless ignored) must be covered by unit tests.
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:511]:
```suggestion
This requires that each PR has a patch coverage of 90%, meaning that 90% of any source added to the repository (unless ignored) must be covered by unit tests.
```

File: doc/source/overview/dev_practices.rst [hunk 47]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+.. note::
+   This is only a sample configuration.
+
+Test Driven Development
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:516]:
```suggestion
Test-Driven Development
```

File: doc/source/overview/dev_practices.rst [hunk 48]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+.. code::
+
+   def send_command(command):
+       """Run a command on the server.
+
+       Parameters
+       ----------
+       command : str
+           Command to run on the remote server. Should be in the form of
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:544]: Command to run on the remote server, which should be in the form ?

File: doc/source/overview/dev_practices.rst [hunk 49]
```diff
@@ -194,4 +194,372 @@ should not wait until a minor release. These are the steps for a patch release:
...
+
+
+
+
+
+Files Layout
+~~~~~~~~~~~~
+PyAnsys libraries should use ``unittest`` or ``pytest`` libraries to run individual
+unit tests contained within a ``tests`` directory in the root of the project.  The
+testing specific files of your project should at a minimum include:
```
Comment (PipKat) [inline doc/source/overview/dev_practices.rst:554]:
```suggestion
specific test files for your project should at a minimum include:
```

**Review by PipKat — APPROVED** [review]

File: doc/source/overview/dev_practices.rst (hunk 24 above)
Comment (MaxJPRey) [reply doc/source/overview/dev_practices.rst:329]: We usually use the word "compile time" such as "runtime". I think we can let it the way it is here.

File: doc/source/overview/dev_practices.rst (hunk 37 above)
Comment (MaxJPRey) [reply doc/source/overview/dev_practices.rst:393]: Here it is about the workflow name.

File: doc/source/overview/dev_practices.rst (hunk 8 above)
Comment (MaxJPRey) [reply doc/source/overview/dev_practices.rst:200]: @PipKat Great idea. Otherwise the sentence is a little bit difficult to read. I am doing it.

File: doc/source/overview/dev_practices.rst (hunk 19 above)
Comment (MaxJPRey) [reply doc/source/overview/dev_practices.rst:309]: They are installed with this code snippet but it is done in a separate command prompt.
The code above is part of a file used for packaging.
I think that we can let it this way to show the separation.

**Review by MaxJPRey — APPROVED** [review]

**Comment by akaszynski** [comment]
Pinging @jorgepiloto, please fix https://github.com/pyansys/dev-guide/runs/5187563828?check_suite_focus=true

**Comment by MaxJPRey** [comment]
> Pinging @jorgepiloto, please fix https://github.com/pyansys/dev-guide/runs/5187563828?check_suite_focus=true

@akaszynski It is done.

File: doc/source/documentation_style/deployment.rst [hunk 50]
```diff
@@ -0,0 +1,14 @@
+.. _documentation_deployment:
+
```
Comment (jorgepiloto) [inline doc/source/documentation_style/deployment.rst:2]: I will remove this blank line, so page link applies the same style as in the rest of `*.rst` files.

File: doc/source/documentation_style/index.rst [hunk 51]
```diff
@@ -29,9 +29,12 @@ The documentation for a PyAnsys library should contain:
 * Developer's guide for the library.
 * Link to this developer's guide.
 
+Finally the documentation should be public and hosted via gh-pages, either as a branch named ``gh-pages`` within the library repository or within a ``gh-pages`` branch within ``<library-repository>-docs``.  For more information, see :ref:`documentation_deployment`.
```
Comment (jorgepiloto) [inline doc/source/documentation_style/index.rst:32]: Add comma after `Finally`.

File: doc/source/guidelines/test_practices.rst [hunk 52]
```diff
@@ -0,0 +1,425 @@
...
+       assert "Created 1" in output
+
+Note that this test only validates the command ``"CREATE,1"`` has been
+received, executed, and sent back to the client. It does not validate all
+commands, but nor is it necessary to do this unless there are edge cases
+(e.g. characters that cannot be streamed or dealing with long running
+commands).
+
+
+Testing Framework
```
Comment (seanpearsonuk) [inline doc/source/guidelines/test_practices.rst:152]: Can we kind of mandate using pytest / unittest for all levels of testing including integration testing. Can we officially frown on use of in-house frameworks/platforms here?

File: doc/source/guidelines/test_practices.rst [hunk 53]
```diff
@@ -0,0 +1,425 @@
...
+
+   pip install .[test]
+
+When using ``pytest``, test via::
+
+   pytest
+
+.. note::
+   We recommend that you place the source of your library within the ``src``
+   direction rather than having your Python library source directly within the
```
Comment (PipKat) [inline doc/source/guidelines/test_practices.rst:187]:
```suggestion
   directory rather than having your Python library source directly within the
```

File: doc/source/guidelines/test_practices.rst [hunk 54]
```diff
@@ -0,0 +1,425 @@
...
+
+   try:
+       import matplotlib
+   except ImportError:  # pragma: no cover
+       raise ImportError("Install matplotlib to use this feature.")
+
+.. note::
+   You should only avoid coverage of parts of your library where you cannot
+   reasonably test without an extensive testing suite or setup.  Most methods
+   and classes, including edge cases, can be reasonable tested. Even parts of
```
Comment (PipKat) [inline doc/source/guidelines/test_practices.rst:240]:
```suggestion
   and classes, including edge cases, can be reasonably tested. Even parts of
```

File: doc/source/guidelines/test_practices.rst [hunk 55]
```diff
@@ -0,0 +1,425 @@
...
+.. code:: yaml
+
+     - name: Install the library
+       run: |
+         pip install .
+         pip install -r requirements_test.txt
+
+
+If you are using ``pyproject.toml`` with the `poetry`_ build system, your
+installation step is:
```
Comment (PipKat) [inline doc/source/guidelines/test_practices.rst:331]:
```suggestion
If you are using ``pyproject.toml`` with the `poetry`_ build system, install
with:
```

File: doc/source/guidelines/test_practices.rst [hunk 56]
```diff
@@ -0,0 +1,425 @@
...
+
+.. note::
+   This is only a sample configuration.
+
+
+Files Layout
+~~~~~~~~~~~~
+PyAnsys libraries should use ``unittest`` or ``pytest`` libraries to run
+individual unit tests contained within a ``tests`` directory in the root of the
+project.  The specific test files for your project should at a minimum include:
```
Comment (PipKat) [inline doc/source/guidelines/test_practices.rst:409]:
```suggestion
project. The specific test files for your project should at a minimum include:
```
```suggestion
project.  The specific test files for your project should at a minimum include:
```

File: doc/source/guidelines/test_practices.rst [hunk 57]
```diff
@@ -0,0 +1,425 @@
...
+
+   requirements_tests.py
+   tests/
+     test_<filename>.py
+     conftest.py
+
+**Requirements File**
+The requirements file contains a list of all the libraries that must be
+installed to run ``pytest``.  No assumption should be made regarding the state
+of the virtual
```
Comment (PipKat) [inline doc/source/guidelines/test_practices.rst:421]: This sentence is incomplete.

Comment (MaxJPRey) [reply doc/source/guidelines/test_practices.rst:421]: @PipKat I will handle that.

File: doc/source/guidelines/test_practices.rst (hunk 52 above)
Comment (MaxJPRey) [reply doc/source/guidelines/test_practices.rst:152]: @seanpearsonuk I will investigate if we can officially get rid of the internal testing tools such as ARM. I will check that it is not mandatory by any ISO9... norm to have an easy access to testing history somewhere or anything else.

File: doc/source/guidelines/test_practices.rst (hunk 57 above)
Comment (MaxJPRey) [reply doc/source/guidelines/test_practices.rst]: Done.
